File: the-maze.cpp
// Time Complexity : O(mn)
// Space Complexity : O(mn)
// Did this code successfully run on Leetcode : Yes

#include <vector>
#include <queue>
  
using namespace std;

class Solution {
public:
    bool hasPath(std::vector<std::vector<int>>& maze, std::vector<int>& start, std::vector<int>& destination) {
        int m = maze.size();
        int n = maze[0].size();

        std::queue<std::vector<int>> q;
        std::vector<std::vector<int>> dirs = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};

        q.push(start);
        maze[start[0]][start[1]] = 2;

        while (!q.empty()) {
            std::vector<int> top = q.front();
            q.pop();

            int r = top[0];
            int c = top[1];

            if (r == destination[0] && c == destination[1]) {
                return true;
            }

            for (const auto& dir : dirs) {
                int i = r;
                int j = c;

                // Keep the ball rolling
                while (i >= 0 && j >= 0 && i < m && j < n && (maze[i][j] == 0 || maze[i][j] == 2)) {
                    i += dir[0];
                    j += dir[1];
                }

                // bring back to stop point
                i -= dir[0];
                j -= dir[1];

                if (maze[i][j] != 2) {
                    q.push({i, j});
                    maze[i][j] = 2;
                }
            }
        }

        return false;
    }
};

File: the-maze.js
// Time Complexity : O(M*N)
// Space Complexity : O(M*N)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

/**
 * @param {number[][]} maze
 * @param {number[]} start
 * @param {number[]} destination
 * @return {boolean}
 */
var hasPath = function(maze, start, destination) {
    if (!maze || !start || !destination) return false;
    
    const q = [];
    const dirs = [[0, 1], [1, 0], [-1, 0], [0, -1]];
    const visited = new Array(maze.length);
    for (const [index, slot] of visited.entries()) {
        visited[index] = new Array(maze[0].length).fill(false);
    };
    visited[start[0]][start[1]] = true;
    q.push(start);
    while (q.length > 0) {
        let currPos = q.pop();
        if (currPos[0] == destination[0] && currPos[1] == destination[1]) return true;
        for (const dir of dirs) {
            let dX = currPos[0] + dir[0];
            let dY = currPos[1] + dir[1];
            while (dX >= 0 && dX < maze.length && dY >= 0 && dY < maze[0].length && maze[dX][dY] == 0) {
                dX += dir[0];
                dY += dir[1];
            }
            if (!visited[dX - dir[0]][dY - dir[1]]) {
                q.push([dX - dir[0], dY - dir[1]]);
                visited[dX - dir[0]][dY - dir[1]] = true;
            }
        }
    }
    return false;
};
